File: ABC170/D_Not_Divisible/main2.scala

// TLEになってしまう

import java.util.Scanner;

object Main extends App {

  val sc = new Scanner(System.in);

  val n = sc.nextInt();
  val as = Array.fill(n)(sc.nextInt());
  val as1 = as.sorted;
  val as2 = new Array[Int](n);

  @scala.annotation.tailrec
  def sub(as1: Array[Int], as1len: Int, count: Int, buf: Array[Int]): Int = {
    if (as1len == 0) {
      count;
    } else if (as1len == 1) {
      count + 1;
    } else {
      val h = as1(0);
      val p = as1.indexWhere(_ > h, 1);
      if (p < 0) {
        count;
      } else {
        var i2: Int = 0;
        (p until as1len).foreach { i1 =>
          val a = as1(i1);
          if (a % h != 0) {
            buf(i2) = a;
            i2 = i2 + 1;
          }
        }
        if (p > 1) {
          sub(buf, i2, count, as1);
        } else {
          sub(buf, i2, count + 1, as1);
        }
      }
    }
  }

  println(sub(as1, n, 0, as2));
}

File: abc183/e/main.2.java
// TLE

import java.util.Scanner;

class Main {
  public static void main(String[] args) {
    int m = 1_000_000_007;

    var sc = new Scanner(System.in);
    int h = sc.nextInt();
    int w = sc.nextInt();
    var ss = new boolean[h*w];
    for (int i = 0; i < h; i++) {
      var t = sc.next();
      for (int j = 0; j < w; j++) {
        if (t.charAt(j) == '#') ss[w*i+j] = true;
      }
    }

    var table = new int[h*w];
    table[0] = 1;
    for (int i = 0; i < h; i++) {
      for (int j = 0; j < w; j++) {
        int g = w*i+j;
        if (ss[g]) continue;
        if (g == 0) continue;
        long x = 0;
        int k = g;
        for (int i2 = i - 1; i2 >= 0; i2--) {
          k -= w;
          int z = table[k];
          if (z == 0) break;
          x += z;
        }
        k = g;
        for (int j2 = j - 1; j2 >= 0; j2--) {
          k -= 1;
          int z = table[k];
          if (z == 0) break;
          x += z;
        }
        int max = (i > j)? j : i;
        k = g;
        for (int s = 1; s <= max; s++) {
          k -= (w+1);
          int z = table[k];
          if (z == 0) break;
          x += z;
        }
        table[g] = (int)(x % m);
      }
    }
    int answer = table[h*w-1];
    System.out.println(answer);
  }
}
